### DISCUSSION TITLE
rsquared for lmfit

### DISCUSSION THREAD

**Discussion by Anselmoo** [lmfit/lmfit-py#801] [Ideas]
Is there actually still interest in implementing `rsquared` as formally proposed https://github.com/lmfit/lmfit-py/pull/81?

This still might fit there?

https://github.com/lmfit/lmfit-py/blob/e54a44c68f9cb40836b9aca5398436f4f2b20b15/lmfit/minimizer.py#L358-L380

See also: [python-lmfit-how-to-calculate-r-squared](https://stackoverflow.com/questions/22581887/python-lmfit-how-to-calculate-r-squared)

**Comment by newville** [comment]
PR #81 was merged more than 8 years ago.  Are you talking about a comment that was both a) made on a closed PR many months after it was merged and b) made more than 7 years ago?   If so, then I would say that implementing `rsquared` was not ever "formally proposed".   It was *sort of suggested in passing in a way that could not possibly be useful* 7 years ago and basically never brought up again.   

You can certainly bring up the idea of adding some more fit statistics - I don't know that anyone would object to that.

**Reply by Anselmoo** [reply]
@newville thx for quick response.

I was referring to https://github.com/lmfit/lmfit-py/pull/81#issuecomment-68129480

I am still thinking that an `R2` analysis can help to further add confidence in the fitting results. However, I don't wanna duplicate a closed Issue and PR, so I was asking, what your thoughts are?

thx again

**Comment by newville** [comment]
@Anselmoo it would be fine to open an Issue on that and/or make a PR to add that.

**Comment by Anselmoo** [comment]
@newville 

I was taking a deeper look into the source code of `lmfit-minimizer`, and I was wondering if it is possible to get somehow the initial `data` and not just the `residual`?

The reason is that for calculating [`R2`](https://en.wikipedia.org/wiki/Coefficient_of_determination), we need the average value of `data` as proposed by [sklearn.metrics.r2_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)

```python
    numerator = (weight * (y_true - y_pred) ** 2).sum(axis=0, dtype=np.float64)
    denominator = (
        weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
    ).sum(axis=0, dtype=np.float64)
```

https://github.com/scikit-learn/scikit-learn/blob/36958fb240fbe435673a9e3c52e769f01f36bec0/sklearn/metrics/_regression.py#L927-L930


Is there a chance to reconstruct data by using `self.userfcn`? I am a little bit lost.

thx for feedback

**Reply by newville** [reply]
Well, no it is not generally possible to get the "data" in the `minimizer` interface.  That simply does not have a concept of "data".  It has an objective function that converts a set of parameters into a residual whose sum-of-squares is minimized.    That residual could be `(data-model)`, but it might not be. It might include multiple sets of different kinds of "data" (in the sense of "observations" or "measurements").  It could simply not include any data at all -- sort of a "non-linear root finding".   In any case, it does not separately think about "data".

The lmfit.Model interface does include a clearer distinction between "data" and "model" for the common case of "curve fitting", where the residual is `weight*(data-model)`.  Any statistic such as R2 that separates "data" and "model" would need to be done there.

**Reply by Anselmoo** [reply]
> The lmfit.Model interface does include a clearer distinction between "data" and "model" for the common case of "curve fitting", where the residual is `weight*(data-model)`. Any statistic such as R2 that separates "data" and "model" would need to be done there.

thx for the input, this sounds very promising. I will take a look
